Support complex numbers in DTensor redistribute

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157329

Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

wconstab committed:
Support complex numbers in DTensor redistribute

Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

[ghstack-poisoned]

wconstab commented:
@wconstab has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77564148).

wconstab committed:
Update on "Support complex numbers in DTensor redistribute"


Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

Differential Revision: [D77564148](https://our.internmc.facebook.com/intern/diff/D77564148)

[ghstack-poisoned]

wconstab committed:
Update on "Support complex numbers in DTensor redistribute"


Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

Differential Revision: [D77564148](https://our.internmc.facebook.com/intern/diff/D77564148)

[ghstack-poisoned]

wconstab commented on test/distributed/tensor/test_redistribute.py:
```diff
@@ -116,6 +127,8 @@ def test_replicate_to_local_partial_grad(self):
             )
             out.backward(torch.ones_like(out))
 
+        # TODO(whc) it appears complex-allreduce is already being supported becuase this test passes,
+        # but I did not see where the support is
```
ah, thanks for pointing that out. That was the last piece i was missing. 

The new piece of info was that CI failed this test for GLOO bc of gloo not supporting allreduce.  I therefore added the complex support at Functional.cpp for allreduce, fixing the gloo test, and now I understand why nccl was already passing without it.  let me remove this TODO.

wconstab committed:
Update on "Support complex numbers in DTensor redistribute"


Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k


[ghstack-poisoned]

XilunWu reviewed (commented):
The test change LGTM but the functional collective part needs small change. Thanks Will for adding complex number support!

wconstab committed:
Update on "Support complex numbers in DTensor redistribute"


Add complex number unwrapping in functional collectives used by DTensor.

Complex tensors are not directly supported by underlying comm kernels
(e.g. nccl) but complex tensors can be viewed as real tensors of a
higher rank (added size-2 tensor dim represents real vs im component).
Collective output is then viewed as complex to restore the
original/expected shape and dtype.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k


[ghstack-poisoned]

XilunWu commented on test/distributed/tensor/test_redistribute.py:
```diff
@@ -581,6 +600,9 @@ def test_shard_dim_alltoall(self):
         self.assertEqual(new_tensor.stride(), new_meta_tensor.stride())
 
 
+instantiate_parametrized_tests(RedistributeTest)
```
note to myself: use `instantiate_parametrized_tests` in DTensor tests

XilunWu reviewed (approved):
stamp to unblock. We can address the comment in follow-up PR

wconstab commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -81,8 +81,21 @@ at::Tensor all_reduce(
+    TORCH_CHECK(
+        // TODO - ideally use 'to_reduce_op' helper but it currently errors on
+        // premul_sum
+        reduce_op == "sum" || reduce_op == "avg" || reduce_op == "premul_sum" ||
```
i agree but it is not trivial.
1) convert string to enum: there is a helper in this file, but it is not complete. if i completed it by filling in missing premul_sum, i would be affecting the behavior of other ops using this helper
2) then i could refactor the complexViewAsRealAllowed out of ProcessGroupNccl.cpp and make it a util and use it.

Happy to do this in another PR, lmk if you have thoughts about (1)

XilunWu replied:
Low prio though. 

> lmk if you have thoughts about (1)

I think you're right. And I don't have a good solution either that can ensure consistency among helpers and potential extension of ReduceOp enum.

wconstab commented:
@pytorchbot merge -i
